### PR TITLE
Improve `clone_from` impls for collections

### PR DESCRIPTION
**Pull request by apasel422**
r? @bluss

### CODE REVIEW

**Comment by dotdash** [comment]
@bors r+

Nice!

**Comment by bluss** [comment]
Did you push a new commit after bors tested the old one. What was the change?

File: src/libcollections/vec.rs [hunk 1]
```diff
@@ -1007,19 +1007,22 @@ impl<T:Clone> Clone for Vec<T> {
...
-            self.truncate(other.len())
-        }
+        self.truncate(other.len());
+        let len = self.len();
 
         // reuse the contained values' allocations/resources.
-        for (place, thing) in self.iter_mut().zip(other) {
-            place.clone_from(thing)
+        {
+            let src = &other[..len];
```
Comment (bluss) [inline src/libcollections/vec.rs]: This trick might warrant a comment.

**Comment by apasel422** [comment]
@bluss Yeah, I pushed a new commit, so bors hasn't tested the new one yet. I removed an unnecessary call to `cmp::min`, since `truncate` guarantees that the destination's length is less than or equal to the source's.

I'll add a comment that explains the trick and re-push.

**Comment by bluss** [comment]
Great

**Comment by apasel422** [comment]
Hmm, instead of using the trick, should I just use a more explicit (and less finicky) `get_unchecked` loop like @Veedrac did in the related issue?

**Comment by bluss** [comment]
It's almost about time we package up this "zip slices" pattern now in a function or iterator actually, since we have to use it in a lot of places.

I think your question is interesting:

1. Write direct, do what I mean, unsafe code  (`get_unchecked`)
2. Write a particular pattern that we know the optimizer understands

(2) can absolutely be called fragile. I still like (2) better since it's safe, it degrades well (still correct even if the optimizer's behavior changes).

**Comment by apasel422** [comment]
It's actually already present in `clone_from_slice`, from what I can tell.

**Comment by bluss** [comment]
yes, for clone in particular it seems it is. We can reuse that here, then.

**Comment by apasel422** [comment]
Updated.

**Comment by bluss** [comment]
Now it's beautiful

@bors r+

**Comment by apasel422** [comment]
Fixed the test failure.

**Comment by bluss** [comment]
@bors r+

thanks!

**Pull request merged**